### ISSUE TITLE
env.es6=false does not disable es6-specific rules

### ISSUE THREAD

**Issue by jokeyrhyme** [eslint/eslint#3686]
.eslintrc:
```json
{
  "env": {
    "es6": false
  },
  "rules": {
    "no-var": 2
  }
}
```

index.js:
```js
'use strict';

var test = 1;

global.console.log(test);
```

$ eslint .
> index.js
>   3:1  error  Unexpected var, use let or const instead  no-var

It would be desirable to be able to maintain a single shared config that enables ES6 rules such as "no-var", and use this config with both ES5 and ES6 projects. Ideally, the only difference in ESLint configuration between such projects would be the value of `env.es6`.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by jokeyrhyme** [comment]
I'm using ESLint v1.3.1.

I think the long term solution is to actually generate the rules documentation by parsing each rule's implementation JavaScript. This means each rule declares which group of rules it belongs in (e.g. "style", "es6", etc). From this, ESLint can identify all the ES6 rules and disable them when `env.es6` is `false`.

A short-term solution would be to just explicitly list all the ES6 rules in the code responsible for disabling them. We'd need a strategy for keeping the list up to date, perhaps a unit test that parses the documentation to identify "es6" rules?

**Comment by jokeyrhyme** [comment]
I think I've identified a possible problem with my proposal. I'm basically arguing for the ability for an environment setting to thoroughly disable a collection of rules, ignoring everywhere else those rules are enabled. Although, maybe this is a special case, where we absolutely want to ignore a user's configuration?

From the documentation, I've identified the following ES6 / ES2015 rules that should not be accidentally enabled when working with ES3 or ES5 codebases:

- [no-dupe-class-members](http://eslint.org/docs/rules/no-dupe-class-members)
- [no-var](http://eslint.org/docs/rules/no-var)
- [object-shorthand](http://eslint.org/docs/rules/object-shorthand)
- [prefer-arrow-callback](http://eslint.org/docs/rules/prefer-arrow-callback)
- [prefer-spread](http://eslint.org/docs/rules/prefer-spread)
- [prefer-reflect](http://eslint.org/docs/rules/prefer-reflect)
- [prefer-template](http://eslint.org/docs/rules/prefer-template)

**Comment by nzakas** [comment]
Environments can't define which rules are being run. That some rules don't apply to a specific environment just means you need to be careful about which rules you enable. Even if you accidentally disable the rules in a non-ES6 environment, it should be very obvious when you run ESLint, so you can disable them.

**Comment by jokeyrhyme** [comment]
I'm working on a PR with an alternative approach. Rules have access to your ESLint settings when they are called, so how about if the rule detects the appropriate "es6" or "ecmaFeature" settings (or lack thereof) and returns a no-op?

**Comment by IanVS** [comment]
@jokeyrhyme Personally I wouldn't like that approach, since it could result in a situation where one believes their code is being linted and passing, but in reality the rule is not being executed due to the environment.  I'd much rather get an error and either disable the rules for that file or enable the correct ecmaFeatures.

**Comment by jokeyrhyme** [comment]
Without the rule disabling itself in situations where it is dangerous (e.g. enforcing ES2015 approaches when code is not ES2015) we'll be forced to maintain multiple shared configs: one for pre-ES2015 code, and one for ES2015.

So people in my position will need to have eslint-config-es5, and eslint-config-es6 (extending -es5). This is definitely explicit, but it's not very convenient for maintaining shared configs, and it means we can't we the `"extends": "@scope"` shortcut for simplicity.

> it could result in a situation where one believes their code is being linted and passing, but in reality the rule is not being executed due to the environment

This isn't a problem with my new approach, because this only affects rules that depend on new ES2015 syntax. If the user has not configured ESLint to enable the new syntax, and they use one of these rules, then they'll get syntax errors. So either the syntax will fail, or the rule will execute. There is no case where the syntax will pass and the rule will noop.

**Comment by jokeyrhyme** [comment]
Regarding syntax changes, I just realised that the "prefer-reflect" rule is really an API change, not a syntax change. As such, there isn't an "ecmaFeatures" flag for it. I completely agree in this case that even if these was such a flag, it could lead to ambiguous situations for users.

**Comment by jokeyrhyme** [comment]
It would be terrific if people who maintain well-known shared configs could weigh in with their thoughts here. @feross? @flet?

**Comment by nzakas** [comment]
@jokeyrhyme I'm sorry, the team doesn't believe that this is a good idea. Having rules enabled and doing nothing is very confusing to end users. We have a policy against anything failing silently, so we're not going to pursue this any further.

I'd suggest looking at using `extends` to help you come up with good configs for individual files.

**Comment by niftylettuce** [comment]
I have the issue that `/*eslint-env es6:false*/` still gives me errors such as `Use let or const instead` if I were to have something like `var path = require('path');` below it.  In the meanwhile I'm having to use `/*eslint-disable */` since I'm not sure what the issue is.

**Comment by ilyavolodin** [comment]
Environments do not control which rules are on or off. If you are getting `Use let or const instead` error, you should disable `no-var` rule.

**Comment by d0f** [comment]
1:1   error    Unexpected var, use let or const instead                  no-var
   2:1   error    Unexpected var, use let or const instead                  no-var
   3:1   error    Unexpected var, use let or const instead                  no-var
   3:5   error    'favicon' is defined but never used                       no-unused-vars
   4:1   error    Unexpected var, use let or const instead                  no-var
   5:1   error    Unexpected var, use let or const instead                  no-var
   6:1   error    Unexpected var, use let or const instead                  no-var
   8:1   error    Unexpected var, use let or const instead                  no-var
   9:1   error    Unexpected var, use let or const instead                  no-var
  11:1   error    Unexpected var, use let or const instead                  no-var

**Comment by cancerberoSgx** [comment]
The easy - manual way I found i just extend the desired shareable config and then turning off all ecma6 rules, for example: 

`// this is just airbnb-base with all es6 specific rules disables

module.exports = {
    "extends": "airbnb-base",
    "rules": {
        "arrow-body-style": "off",
        "arrow-parens": "off",
        "arrow-spacing": "off",
        "constructor-super": "off",
        "generator-star-spacing": "off",
        "no-class-assign": "off",
        "no-confusing-arrow": "off",
        "no-const-assign": "off",
        "no-dupe-class-members": "off",
        "no-duplicate-imports": "off",
        "no-new-symbol": "off",
        "no-restricted-imports": "off",
        "no-this-before-super": "off",
        "no-useless-computed-key": "off",
        "no-useless-constructor": "off",
        "no-useless-rename": "off",
        "no-var": "off",
        "object-shorthand": "off",
        "prefer-arrow-callback": "off",
        "prefer-const": "off",
        "prefer-destructuring": "off",
        "prefer-numeric-literals": "off",
        "prefer-rest-params": "off",
        "prefer-spread": "off",
        "prefer-template": "off",
        "require-yield": "off",
        "rest-spread-spacing": "off",
        "sort-imports": "off",
        "symbol-description": "off",
        "template-curly-spacing": "off",
        "yield-star-spacing": "off"
    }
};`

The bad thing is that I need to manually maintain this list if new es6 rules are defined - but is reasonable I think. Perhaps the feature missing in eslint is be able to group rules with names - probably the groups are the different sections defined https://eslint.org/docs/rules/ - this way i can easily disable all using just  one meta-rule
